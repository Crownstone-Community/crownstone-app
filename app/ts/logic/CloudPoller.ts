import {LOGe} from "../logging/Log";
import {NotificationHandler} from "../backgroundProcesses/NotificationHandler";
import {InviteCenter} from "../backgroundProcesses/InviteCenter";

class CloudPollerClass {

  poll(forSync: boolean = false) : Promise<void> {
    if (!NotificationHandler.notificationPermissionGranted || forSync) {
      return new Promise<void>((resolve, reject) => { resolve(); })
        .then(() => { return InviteCenter.checkForInvites();   })
        .catch((err) => {
          LOGe.cloud("CloudPoller: Failed to poll.", err?.message);
        })
    }
    else {
      return Promise.resolve();
    }
  }
}

export const CloudPoller = new CloudPollerClass();
