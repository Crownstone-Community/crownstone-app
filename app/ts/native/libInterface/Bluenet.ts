import { NativeModules } from 'react-native';
import { DISABLE_NATIVE } from '../../ExternalConfig'
import { LOGi } from "../../logging/Log";

export let Bluenet;

const BluenetAPI = {
  clearTrackedBeacons:      () => { console.log("BluenetBridgeCall: clearTrackedBeacons: "); },        // called through BluenetPromiseWrapper --> must be promise.
  rerouteEvents:            () => { console.log("BluenetBridgeCall: rerouteEvents:       "); },
  isReady:                  () => { console.log("BluenetBridgeCall: isReady:             "); },                    // called through BluenetPromiseWrapper --> must be promise.
  connect:                  () => { console.log("BluenetBridgeCall: connect:             "); },                    // called through BluenetPromiseWrapper --> must be promise.
  disconnectCommand:        () => { console.log("BluenetBridgeCall: disconnectCommand:   "); },          // called through BluenetPromiseWrapper --> must be promise.
  phoneDisconnect:          () => { console.log("BluenetBridgeCall: phoneDisconnect:     "); },            // called through BluenetPromiseWrapper --> must be promise.
  resetBle:                 () => { console.log("BluenetBridgeCall: resetBle:            "); },
  startScanning:            () => { console.log("BluenetBridgeCall: startScanning:       "); },
  startScanningForCrownstones:           () => { console.log("BluenetBridgeCall: startScanningForCrownstones:           "); },
  startScanningForCrownstonesUniqueOnly: () => { console.log("BluenetBridgeCall: startScanningForCrownstonesUniqueOnly: "); },
  stopScanning:             () => { console.log("BluenetBridgeCall: stopDFU:    "); },
  requestBleState:          () => { console.log("BluenetBridgeCall: requestBleState: "); },// Send events "bleStatus" and "locationStatus" with the current state.

  startIndoorLocalization:  () => { console.log("BluenetBridgeCall: startIndoorLocalization: "); },
  stopIndoorLocalization:   () => { console.log("BluenetBridgeCall: stopIndoorLocalization:  "); },

  requestLocation:          () => { console.log("BluenetBridgeCall: requestLocation:          "); },// Should return data {"latitude": number, "longitude": number}. Called through BluenetPromiseWrapper --> must be promise.
  requestLocationPermission:() => { console.log("BluenetBridgeCall: requestLocationPermission:"); },// Request for location permission during tutorial.
  trackIBeacon:             () => { console.log("BluenetBridgeCall: trackIBeacon:             "); },// Add the UUID to the list of tracked iBeacons, associate it with given sphereId, and start tracking.
  stopTrackingIBeacon:      () => { console.log("BluenetBridgeCall: stopTrackingIBeacon:      "); },// Remove the UUID from the list of tracked iBeacons.
  pauseTracking:            () => { console.log("BluenetBridgeCall: pauseTracking:            "); },// Stop tracking, but keep the list of tracked iBeacon UUIDs. Stop sending any tracking events: iBeacon, enter/exit region. Assume all tracked iBeacon UUIDs are out the region.
  resumeTracking:           () => { console.log("BluenetBridgeCall: resumeTracking:           "); },// Start tracking again, with the list that is already there.

  startCollectingFingerprint:  () => { console.log("BluenetBridgeCall: startCollectingFingerprint:  "); },
  abortCollectingFingerprint:  () => { console.log("BluenetBridgeCall: abortCollectingFingerprint:  "); },
  pauseCollectingFingerprint:  () => { console.log("BluenetBridgeCall: pauseCollectingFingerprint:  "); },
  resumeCollectingFingerprint: () => { console.log("BluenetBridgeCall: resumeCollectingFingerprint: "); },
  finalizeFingerprint:         () => { console.log("BluenetBridgeCall: finalizeFingerprint:         "); },       // called through BluenetPromiseWrapper --> must be promise. Promise return value is a stringified fingerprint

  loadFingerprint:          () => { console.log("BluenetBridgeCall: loadFingerprint:       "); },
  getMACAddress:            () => { console.log("BluenetBridgeCall: getMACAddress:         "); },             // called through BluenetPromiseWrapper --> must be promise.
  commandFactoryReset:      () => { console.log("BluenetBridgeCall: commandFactoryReset:   "); },       // called through BluenetPromiseWrapper --> must be promise.
  recover:                  () => { console.log("BluenetBridgeCall: recover:               "); },                   // called through BluenetPromiseWrapper --> must be promise.
  setupCrownstone:          () => { console.log("BluenetBridgeCall: setupCrownstone:       "); },           // called through SetupCrownstone in BLEUtil
  quitApp:                  () => { console.log("BluenetBridgeCall: quitApp:               "); NativeModules.BluenetJS.quitApp() },                   // Used to quit the app during logout
  enableLoggingToFile:      () => { console.log("BluenetBridgeCall: enableLoggingToFile:   "); },
  enableExtendedLogging:    () => { console.log("BluenetBridgeCall: enableExtendedLogging: "); },
  clearLogs:                () => { console.log("BluenetBridgeCall: clearLogs:             "); },
  crash:                    () => { console.log("BluenetBridgeCall: crash!                 "); },

  // mesh
  multiSwitch:              () => { console.log("BluenetBridgeCall: multiSwitch:             "); },
  getHardwareVersion:       () => { console.log("BluenetBridgeCall: getHardwareVersion:      "); },
  getBootloaderVersion:     () => { console.log("BluenetBridgeCall: getBootloaderVersion:    "); },
  getFirmwareVersion:       () => { console.log("BluenetBridgeCall: getFirmwareVersion:      "); },
  getUICR:                  () => { console.log("BluenetBridgeCall: geUICR:                  "); },
  bootloaderToNormalMode:   () => { console.log("BluenetBridgeCall: bootloaderToNormalMode:  "); },
  clearFingerprintsPromise: () => { console.log("BluenetBridgeCall: clearFingerprintsPromise:"); },
  clearFingerprints:        () => { console.log("BluenetBridgeCall: clearFingerprints:       "); },
  setTime:                  () => { console.log("BluenetBridgeCall: setTime:                 "); },
  meshSetTime:              () => { console.log("BluenetBridgeCall: meshSetTime:             "); },
  batterySaving:            () => { console.log("BluenetBridgeCall: batterySaving:           "); },
  setBackgroundScanning:    () => { console.log("BluenetBridgeCall: setBackgroundScanning:   "); },

  viewsInitialized:         () => { console.log("BluenetBridgeCall: viewsInitialized:  "); },
  lockSwitch:               () => { console.log("BluenetBridgeCall: lockSwitch:        "); },
  allowDimming:             () => { console.log("BluenetBridgeCall: allowDimming:      "); },
  setSwitchCraft:           () => { console.log("BluenetBridgeCall: setSwitchCraft:    "); },
  sendNoOp:                 () => { console.log("BluenetBridgeCall: sendNoOp:          "); },
  sendMeshNoOp:             () => { console.log("BluenetBridgeCall: sendMeshNoOp:      "); },


  getSwitchState:           () => { console.log("BluenetBridgeCall:  getSwitchState:   "); },
  putInDFU:                 () => { console.log("BluenetBridgeCall:  putInDFU:         "); },
  performDFU:               () => { console.log("BluenetBridgeCall:  repair:       "); },
  restartCrownstone:        () => { console.log("BluenetBridgeCall:  restartCrownstone:"); },
  clearKeySets:             () => { console.log("BluenetBridgeCall:  clearKeysets:     "); },
  setKeySets:               () => { console.log("BluenetBridgeCall:  setKeySets:       "); },
  setupFactoryReset:        () => { console.log("BluenetBridgeCall:  setupFactoryReset:"); },
  setupPutInDFU:            () => { console.log("BluenetBridgeCall:  setupPutInDFU:    "); },
  toggleSwitchState:        () => { console.log("BluenetBridgeCall:  toggleSwitchState:"); },
  getTrackingState:         () => { console.log("BluenetBridgeCall:  getTrackingState: "); },
  setDevicePreferences:     () => { console.log("BluenetBridgeCall:  setDevicePreferences: "); },
  setLocationState:         () => { console.log("BluenetBridgeCall:  setLocationState: "); },
  startAdvertising:         () => { console.log("BluenetBridgeCall:  startAdvertising: "); },
  stopAdvertising:          () => { console.log("BluenetBridgeCall:  stopAdvertising: "); },

  setCrownstoneNames:       () => { console.log("BluenetBridgeCall:  setCrownstoneNames: "); },
  setupPulse:               () => { console.log("BluenetBridgeCall:  setupPulse: "); },

  subscribeToNearest:       () => { console.log("BluenetBridgeCall:  subscribeToNearest "); },
  unsubscribeNearest:       () => { console.log("BluenetBridgeCall:  unsubscribeNearest "); },
  subscribeToUnverified:    () => { console.log("BluenetBridgeCall:  subscribeToUnverified "); },
  unsubscribeUnverified:    () => { console.log("BluenetBridgeCall:  unsubscribeUnverified "); },

  initBroadcasting:            () => { console.log("BluenetBridgeCall:  initBroadcasting "); },
  checkBroadcastAuthorization: () => { console.log("BluenetBridgeCall:  checkBroadcastAuthorization "); },

  isPeripheralReady:           () => { console.log("BluenetBridgeCall: isPeripheralReady"); },
  setSwitchState:              () => { console.log("BluenetBridgeCall: setSwitchState"); },
  isDevelopmentEnvironment:    () => { console.log("BluenetBridgeCall: isDevelopmentEnvironment"); },
  clearErrors:                 () => { console.log("BluenetBridgeCall: clearErrors"); },
  broadcastSwitch:             () => { console.log("BluenetBridgeCall: broadcastSwitch"); },
  addBehaviour:                () => { console.log("BluenetBridgeCall: addBehaviour"); },
  updateBehaviour:             () => { console.log("BluenetBridgeCall: updateBehaviour"); },
  removeBehaviour:             () => { console.log("BluenetBridgeCall: removeBehaviour"); },
  getBehaviour:                () => { console.log("BluenetBridgeCall: getBehaviour"); },
  switchRelay:                 () => { console.log("BluenetBridgeCall: switchRelay"); },
  switchDimmer:                () => { console.log("BluenetBridgeCall: switchDimmer"); },
  getResetCounter:             () => { console.log("BluenetBridgeCall: getResetCounter"); },
  getSwitchcraftThreshold:     () => { console.log("BluenetBridgeCall: getSwitchcraftThreshold"); },
  getMaxChipTemp:              () => { console.log("BluenetBridgeCall: getMaxChipTemp"); },
  getDimmerCurrentThreshold:   () => { console.log("BluenetBridgeCall: getDimmerCurrentThreshold"); },
  getDimmerTempUpThreshold:    () => { console.log("BluenetBridgeCall: getDimmerTempUpThreshold"); },
  getDimmerTempDownThreshold:  () => { console.log("BluenetBridgeCall: getDimmerTempDownThreshold"); },
  getVoltageZero:              () => { console.log("BluenetBridgeCall: getVoltageZero"); },
  getCurrentZero:              () => { console.log("BluenetBridgeCall: getCurrentZero"); },
  getPowerZero:                () => { console.log("BluenetBridgeCall: getPowerZero"); },
  getVoltageMultiplier:        () => { console.log("BluenetBridgeCall: getVoltageMultiplier"); },
  getCurrentMultiplier:        () => { console.log("BluenetBridgeCall: getCurrentMultiplier"); },
  setSwitchcraftThreshold:     () => { console.log("BluenetBridgeCall: setSwitchcraftThreshold"); },
  setMaxChipTemp:              () => { console.log("BluenetBridgeCall: setMaxChipTemp"); },
  setDimmerCurrentThreshold:   () => { console.log("BluenetBridgeCall: setDimmerCurrentThreshold"); },
  setDimmerTempUpThreshold:    () => { console.log("BluenetBridgeCall: setDimmerTempUpThreshold"); },
  setDimmerTempDownThreshold:  () => { console.log("BluenetBridgeCall: setDimmerTempDownThreshold"); },
  setVoltageZero:              () => { console.log("BluenetBridgeCall: setVoltageZero"); },
  setCurrentZero:              () => { console.log("BluenetBridgeCall: setCurrentZero"); },
  setPowerZero:                () => { console.log("BluenetBridgeCall: setPowerZero"); },
  setVoltageMultiplier:        () => { console.log("BluenetBridgeCall: setVoltageMultiplier"); },
  setCurrentMultiplier:        () => { console.log("BluenetBridgeCall: setCurrentMultiplier"); },
  setUartState:                () => { console.log("BluenetBridgeCall: setUartState"); },

  setTapToToggle:                      () => { console.log("BluenetBridgeCall: setTapToToggle"); },
  setTapToToggleThresholdOffset:       () => { console.log("BluenetBridgeCall: setTapToToggleThresholdOffset"); },


  setSunTimes:                         () => { console.log("BluenetBridgeCall: setSunTimes"); },
  setSunTimesViaConnection:            () => { console.log("BluenetBridgeCall: setSunTimesViaConnection"); },
  turnOnMesh:                          () => { console.log("BluenetBridgeCall: turnOnMesh"); },
  turnOnBroadcast:                     () => { console.log("BluenetBridgeCall: turnOnBroadcast"); },
  broadcastBehaviourSettings:          () => { console.log("BluenetBridgeCall: broadcastBehaviourSettings"); },


  syncBehaviours:                      () => { console.log("BluenetBridgeCall: syncBehaviours"); },
  getBehaviourMasterHash:              () => { console.log("BluenetBridgeCall: getBehaviourMasterHash"); },
  setTimeViaBroadcast:                 () => { console.log("BluenetBridgeCall: setTimeViaBroadcast"); },
  broadcastExecute:                    () => { console.log("BluenetBridgeCall: broadcastExecute"); },


  transferHubTokenAndCloudId:          () => { console.log("BluenetBridgeCall: transferHubTokenAndCloudId"); },
  setUartKey:                          () => { console.log("BluenetBridgeCall: setUartKey"); },
  requestCloudId:                      () => { console.log("BluenetBridgeCall: requestCloudId"); },
  factoryResetHub:                     () => { console.log("BluenetBridgeCall: factoryResetHub"); },
  factoryResetHubOnly:                 () => { console.log("BluenetBridgeCall: factoryResetHubOnly"); },

  cancelConnectionRequest:             () => { console.log("BluenetBridgeCall: cancelConnectionRequest"); },
  getBehaviourDebugInformation:        () => { console.log("BluenetBridgeCall: getBehaviourDebugInformation"); },
  getTapToToggleThresholdOffset:       () => { console.log("BluenetBridgeCall: getTapToToggleThresholdOffset"); },
  canUseDynamicBackgroundBroadcasts:   () => { console.log("BluenetBridgeCall: canUseDynamicBackgroundBroadcasts"); },
  registerTrackedDevice:               () => { console.log("BluenetBridgeCall: registerTrackedDevice"); },
  trackedDeviceHeartbeat:              () => { console.log("BluenetBridgeCall: trackedDeviceHeartbeat"); },
  broadcastUpdateTrackedDevice:        () => { console.log("BluenetBridgeCall: broadcastUpdateTrackedDevice"); },
  getCrownstoneUptime:                 () => { console.log("BluenetBridgeCall: getCrownstoneUptime"); },
  getAdcRestarts:                      () => { console.log("BluenetBridgeCall: getAdcRestarts"); },
  getSwitchHistory:                    () => { console.log("BluenetBridgeCall: getSwitchHistory"); },
  getPowerSamples:                     () => { console.log("BluenetBridgeCall: getPowerSamples"); },
  getMinSchedulerFreeSpace:            () => { console.log("BluenetBridgeCall: getMinSchedulerFreeSpace"); },
  getLastResetReason:                  () => { console.log("BluenetBridgeCall: getLastResetReason"); },
  getGPREGRET:                         () => { console.log("BluenetBridgeCall: getGPREGRET"); },
  getAdcChannelSwaps:                  () => { console.log("BluenetBridgeCall: getAdcChannelSwaps"); },
  setSoftOnSpeed:                      () => { console.log("BluenetBridgeCall: setSoftOnSpeed"); },
  getSoftOnSpeed:                      () => { console.log("BluenetBridgeCall: getSoftOnSpeed"); },


  getConstants:                        () => { console.log("BluenetBridgeCall: getConstants"); },

  addListener:                         () => { console.log("BluenetBridgeCall: addListener"); },
  removeListeners:                     () => { console.log("BluenetBridgeCall: removeListener"); },

};

if (DISABLE_NATIVE === true) {
  // LOG.info("!----------- --- --- --- -- -- -- - - - -- -- -- --- --- --- -----------!");
  // LOG.info("!-----------  NATIVE CALLS ARE DISABLED BY EXTERNALCONFIG.JS -----------!");
  // LOG.info("!----------- --- --- --- -- -- -- - - - -- -- -- --- --- --- -----------!");
  Bluenet = BluenetAPI;
}
else if (NativeModules.BluenetJS) {
  if (global.__DEV__ ) {
    let wrappedBluenet = {};
    Object.keys(NativeModules.BluenetJS).forEach((key) => {
      wrappedBluenet[key] = function(param, param2, param3, param4, param5) {
        let bluenetArguments = [];
        // @ts-ignore
        for (let i = 0; i < arguments.length; i++) {
          // @ts-ignore
          if (arguments[i] !== undefined) {
            // @ts-ignore
            bluenetArguments.push(arguments[i])
          }
        }
        if (key === "loadFingerprint") {
          LOGi.info("BLUENET CALL:", key, '<redacted>');
        }
        else {
          LOGi.info("BLUENET CALL:", key, bluenetArguments);
        }
        return NativeModules.BluenetJS[key].apply(this, bluenetArguments);
      }
    })
    Bluenet = wrappedBluenet;
  }
  else {
    Bluenet = NativeModules.BluenetJS;
  }

  let API_Keys = Object.keys(BluenetAPI);
  let notImplemented = [];
  for (let i = 0; i < API_Keys.length; i++) {
    if (Bluenet[API_Keys[i]] === undefined) {
      notImplemented.push(API_Keys[i])
    }
  }
  if (notImplemented.length > 0) {
    console.error("Unimplemented method in Bridge file:", notImplemented);
  }

  let existingKeys = Object.keys(Bluenet);
  let extraMethods = [];
  for (let i = 0; i < existingKeys.length; i++) {
    if (BluenetAPI[existingKeys[i]] === undefined) {
      extraMethods.push(existingKeys[i])
    }
  }
  if (extraMethods.length > 0) {
    console.warn("Additional methods detected in Bridge file:", extraMethods);
  }
}
