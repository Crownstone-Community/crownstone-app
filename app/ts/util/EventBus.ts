import { LOGd, LOGe, LOGi} from "../logging/Log";
import { xUtil } from "./StandAloneUtil";


const EXCLUDE_FROM_CLEAR = {
  showLoading: true,
};

export class EventBusClass {
  _id: string;
  _topics : object;
  _topicIds : object;

  _count = 0
  _type : string;

  constructor(type: string) {
    this._type = "EventBus_" + type;
    this._id = xUtil.getUUID();
    this._topics = {};
    this._topicIds = {};
  }
  
  on(topic, callback) {
    if (!(topic)) {
      LOGe.event('Attempting to subscribe to undefined topic:', topic);
      return;
    }
    if (!(callback)) {
      LOGe.event('Attempting to subscribe without callback to topic:', topic);
      return;
    }

    if (this._topics[topic] === undefined) {
      this._topics[topic] = [];
    }

    // generate unique id
    let id = xUtil.getUUID();

    this._topics[topic].push({id,callback});
    this._topicIds[id] = true;
    this._count += 1;
    LOGi.event(`Subscribed to topic[${topic}], topicCount:[${this._topics[topic].length}], totalCount:[${this._count}] type:[${this._type}] busId:[${this._id}]`);

    // return unsubscribe function.
    return () => {
      if (this._topics[topic] !== undefined) {
        // find id and delete
        for (let i = 0; i < this._topics[topic].length; i++) {
          if (this._topics[topic][i].id === id) {
            this._topics[topic].splice(i,1);
            this._count -= 1;
            LOGi.event(`Unsubscribed from topic[${topic}], topicCount:[${this._topics[topic].length}], totalCount:[${this._count}] type:[${this._type}] busId:[${this._id}]`);
            break;
          }
        }

        // clear the ID
        this._topicIds[id] = undefined;
        delete this._topicIds[id];

        if (this._topics[topic].length === 0) {
          delete this._topics[topic];
        }
      }
    };
  }

  emit(topic, data?) {
    if (this._topics[topic] !== undefined) {
      LOGd.event(topic, data);
      // Firing these elements can lead to a removal of a point in this._topics.
      // To ensure we do not cause a shift by deletion (thus skipping a callback) we first put them in a separate Array
      let fireElements = [];

      for (let i = 0; i < this._topics[topic].length; i++) {
        fireElements.push(this._topics[topic][i]);
      }

      for (let i = 0; i < fireElements.length; i++) {
        // this check makes sure that if a callback has been deleted, we do not fire it.
        if (this._topicIds[fireElements[i].id] === true) {
          fireElements[i].callback(data);
        }
      }
    }
  }

  /**
   * Emit after a tick to get let the calling function finish first
   * @param topic
   * @param callback
   */
  emitAfterTick(topic, callback) {
    setTimeout(() => {
      this.emit(topic, callback);
    }, 0);
  }

  once(topic, callback) {
    let unsubscriber = this.on(topic, (data: any) => {
      unsubscriber();
      callback(data);
    });
    return unsubscriber;
  }


  clearAllEvents() {
    this._count = 0;
    LOGi.event(`EventBus: Clearing all event listeners type:[${this._type}] busId:[${this._id}]`);
    this._topics = {};
    this._topicIds = {};
  }


  /**
   * This will only be used at clearing the database.
   */
  clearMostEvents() {
    let topics = Object.keys(this._topics);
    let remainingTopics = {};
    this._count = 0;
    for (let i = 0; i < topics.length; i++) {
      if (EXCLUDE_FROM_CLEAR[topics[i]] !== true) {
        delete this._topics[topics[i]];
      }
      else {
        remainingTopics[topics[i]] = this._topics[topics[i]].length;
        this._count += this._topics[topics[i]].length
      }
    }

    for (let topic in remainingTopics) {
      LOGi.event(`EventBus: RemainingTopics topic[${topic}], topicCount:[${remainingTopics[topic].length}], totalCount:[${this._count}] type:[${this._type}] busId:[${this._id}]`);
    }


    LOGi.event(`EventBus: Clearing most event listeners. totalCount:[${this._count}] type:[${this._type}] busId:[${this._id}]`);

  }
}

export let eventBus : any = new EventBusClass("mainSingleton");
