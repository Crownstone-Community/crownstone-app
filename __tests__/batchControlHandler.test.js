// __tests__/Intro-test.js
// Note: test renderer must be required after react-native.
let jest = require("jest");
jest.mock('react-native-fs', () => {return {};});
jest.mock('react-native-device-info');

jest.mock('../js/ExternalConfig', () => {
  return {
    CLOUD_ADDRESS: 'https://crownstone-cloud-dev.herokuapp.com/api/',
    DEBUG: true,
    LOG_SCHEDULER: false,
    LOG_BLE: false,
    LOG_EVENTS: true,
    LOG_STORE: false,
    LOG_MESH: true,
    LOG_CLOUD: true,
    LOG_DEBUG: true,
    LOGGING: true,
    LOG_ERRORS: true,
    LOG_WARNINGS: true,
    LOG_VERBOSE: true,
    LOG_TO_FILE: false,
    DISABLE_NATIVE: false,
    SILENCE_CLOUD: true,
    OVERRIDE_DATABASE: false,
    NO_LOCATION_NAME: 'None',
    ENCRYPTION_ENABLED: true,
    AMOUNT_OF_CROWNSTONES_FOR_INDOOR_LOCALIZATION: 4,
    NETWORK_REQUEST_TIMEOUT: 15000,
    HIGH_FREQUENCY_SCAN_MAX_DURATION: 15000,
    DISABLE_TIMEOUT: 30000,
    KEEPALIVE_INTERVAL: 60,
    KEEPALIVE_REPEAT_ATTEMPTS: 1,
    RESET_TIMER_FOR_NEAR_AWAY_EVENTS: 20000,
    RELEASE_MODE: false,
    TESTING_IN_PROCESS: true,
    LOCAL_TESTING: false
  }
});

import { eventBus } from "../js/util/EventBus"

jest.mock('../js/native/Bluenet', () => {
  return {
    eventBus: null,
    success: true,
    Bluenet: {
      connect: (handle, callback) => {
        this.eventBus.emit('test', 'connect', arguments);
        console.log("connecting to ", handle);
        setTimeout(() => {
          callback({error: !this.success});
        },250)
      },
      multiSwitch: (arr, callback) => {
        this.eventBus.emit('test', 'multiSwitch', arguments);
        console.log("multiSwitch to ", arr);
        callback({error: !this.success});
      },
      disconnect: (callback) => {
        this.eventBus.emit('test', 'disconnect', arguments);
        console.log("disconnect");
        callback({error: !this.success});
      },
      phoneDisconnect: (callback) => {
        this.eventBus.emit('test', 'phoneDisconnect', arguments);
        console.log("disconnect");
        callback({error: !this.success});
      },
    },
    loadEventBus: (bus) => {
      this.eventBus = bus;
      bus.on("setState", (success) => { this.success = success; })
    }
  }
});

import * as mockBluenet from "../js/native/Bluenet"
import { EventBusClass } from "../js/util/EventBus"
import { BatchCommandHandler } from "../js/logic/BatchCommandHandler"


test('object assignment', () => {
  // prep
  let testBus = new EventBusClass();
  mockBluenet.loadEventBus(testBus);

  testBus.on('test', (functionName, params) => {console.log("receivedTestEvent", functionName, params);})
  return new Promise((resolve, reject) => {
    let mockStone = {config: {crownstoneId:'iAmFrank', meshNetworkId : 123, handle:'yet to find me'}};
    let keepAlive =      { commandName:'keepAlive'};
    let keepAliveState = { commandName:'keepAliveState', state: 1, timeout: 150, changeState: true };
    let setSwitchState = { commandName:'setSwitchState', state: 1};
    let multiSwitch =    { commandName:'multiSwitch',    state: 1, timeout: 0, intent: 4};

    let meshEmit = {
      handle: 'this is my handle',
      stoneId: 'la stone id',
      meshNetworkId: 123,
      rssi: -60,
    };
    let directEmit = {
      handle: 'this is my handle',
      stoneId: 'la stone id',
      rssi: -60,
    };

    BatchCommandHandler.load(mockStone, 'stoneId', 'sphereId', multiSwitch).catch((x) => {console.log('load',x)});
    BatchCommandHandler.execute().catch((x) => {console.log('execute',x)});
    let { directCommands, meshNetworks } = BatchCommandHandler._extractTodo();

    console.log(BatchCommandHandler._extractTodo().meshNetworks.sphereId['123']);

    eventBus.emit("updateMeshNetwork_sphereId_123", meshEmit);

    expect({one: 1, two: 2}).toEqual({one: 1, two: 2});


    setTimeout(() => {resolve()}, 500)
  })
});