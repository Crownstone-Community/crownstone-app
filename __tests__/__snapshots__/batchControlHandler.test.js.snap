// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchCommandHandler Mesh 1`] = `
Object {
  "sphereId": Array [],
}
`;

exports[`BatchCommandHandler Mesh 2`] = `
Object {
  "sphereId": Object {
    "1": Object {
      "keepAlive": Array [],
      "keepAliveState": Array [],
      "multiSwitch": Array [
        Object {
          "attempts": 1,
          "cleanup": [Function],
          "crownstoneId": "CSID1",
          "handle": "handle-CSID-1",
          "intent": 4,
          "promise": Object {
            "pending": false,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 0,
        },
        Object {
          "attempts": 1,
          "cleanup": [Function],
          "crownstoneId": "CSID2",
          "handle": "handle-CSID-2",
          "intent": 4,
          "promise": Object {
            "pending": false,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 0,
        },
        Object {
          "attempts": 1,
          "cleanup": [Function],
          "crownstoneId": "CSID3",
          "handle": "handle-CSID-3",
          "intent": 4,
          "promise": Object {
            "pending": false,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 0,
        },
      ],
      "other": Array [],
      "setSwitchState": Array [],
    },
    "2": Object {
      "keepAlive": Array [],
      "keepAliveState": Array [],
      "multiSwitch": Array [
        Object {
          "attempts": 1,
          "cleanup": [Function],
          "crownstoneId": "CSID4",
          "handle": "handle-CSID-4",
          "intent": 4,
          "promise": Object {
            "pending": false,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 0,
        },
      ],
      "other": Array [],
      "setSwitchState": Array [],
    },
  },
}
`;

exports[`BatchCommandHandler Mesh 3`] = `
Object {
  "sphereId": Array [],
}
`;

exports[`BatchCommandHandler Mesh 4`] = `
Object {
  "sphereId": Object {
    "1": Object {
      "keepAlive": Array [],
      "keepAliveState": Array [
        Object {
          "attempts": 1,
          "changeState": true,
          "cleanup": [Function],
          "crownstoneId": "CSID2",
          "handle": "handle-CSID-2",
          "promise": Object {
            "pending": false,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 150,
        },
        Object {
          "attempts": 1,
          "changeState": true,
          "cleanup": [Function],
          "crownstoneId": "CSID3",
          "handle": "handle-CSID-3",
          "promise": Object {
            "pending": false,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 150,
        },
      ],
      "multiSwitch": Array [
        Object {
          "attempts": 1,
          "cleanup": [Function],
          "crownstoneId": "CSID1",
          "handle": "handle-CSID-1",
          "intent": 4,
          "promise": Object {
            "pending": true,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 0,
        },
        Object {
          "attempts": 1,
          "cleanup": [Function],
          "crownstoneId": "CSID2",
          "handle": "handle-CSID-2",
          "intent": 4,
          "promise": Object {
            "pending": true,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 0,
        },
        Object {
          "attempts": 1,
          "cleanup": [Function],
          "crownstoneId": "CSID3",
          "handle": "handle-CSID-3",
          "intent": 4,
          "promise": Object {
            "pending": true,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 0,
        },
      ],
      "other": Array [],
      "setSwitchState": Array [],
    },
    "2": Object {
      "keepAlive": Array [],
      "keepAliveState": Array [],
      "multiSwitch": Array [
        Object {
          "attempts": 1,
          "cleanup": [Function],
          "crownstoneId": "CSID4",
          "handle": "handle-CSID-4",
          "intent": 4,
          "promise": Object {
            "pending": false,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 0,
        },
      ],
      "other": Array [],
      "setSwitchState": Array [],
    },
  },
}
`;

exports[`BatchCommandHandler Mesh 5`] = `
Object {
  "sphereId": Array [],
}
`;

exports[`BatchCommandHandler Mesh 6`] = `
Object {
  "sphereId": Object {
    "1": Object {
      "keepAlive": Array [],
      "keepAliveState": Array [
        Object {
          "attempts": 1,
          "changeState": true,
          "cleanup": [Function],
          "crownstoneId": "CSID2",
          "handle": "handle-CSID-2",
          "promise": Object {
            "pending": true,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 150,
        },
        Object {
          "attempts": 1,
          "changeState": true,
          "cleanup": [Function],
          "crownstoneId": "CSID3",
          "handle": "handle-CSID-3",
          "promise": Object {
            "pending": true,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 150,
        },
        Object {
          "attempts": 1,
          "changeState": true,
          "cleanup": [Function],
          "crownstoneId": "CSID1",
          "handle": "handle-CSID-1",
          "promise": Object {
            "pending": false,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 150,
        },
        Object {
          "attempts": 1,
          "changeState": true,
          "cleanup": [Function],
          "crownstoneId": "CSID3",
          "handle": "handle-CSID-3",
          "promise": Object {
            "pending": false,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 150,
        },
      ],
      "multiSwitch": Array [],
      "other": Array [],
      "setSwitchState": Array [],
    },
    "2": Object {
      "keepAlive": Array [],
      "keepAliveState": Array [],
      "multiSwitch": Array [
        Object {
          "attempts": 1,
          "cleanup": [Function],
          "crownstoneId": "CSID4",
          "handle": "handle-CSID-4",
          "intent": 4,
          "promise": Object {
            "pending": false,
            "reject": [Function],
            "resolve": [Function],
          },
          "state": 1,
          "timeout": 0,
        },
      ],
      "other": Array [],
      "setSwitchState": Array [],
    },
  },
}
`;
